File: miniprojects/battleship_project/battleship.js
const Board = require('./board');
const Player = require('./player');

class Battleship{
    constructor(n){
        this.player = new Player();
        this.board = new Board(n);
        this.remainingMisses = Math.floor((n * n) / 2);
    }

    startGame(){
        this.board.placeRandomShips();
        const numShips = this.board.numShips();
        console.log(`There are currently ${numShips} hidden ships.`);
        this.board.print();
    }
    lose(){
        if(this.remainingMisses <= 0){
            console.log('you lose');
            return true;
        }else{
            return false;
        }
    }
    win(){
        if(this.board.numShips() === 0){
            console.log('you win.');
            return true;
        }else{
            return false;
        }
    }
    gameOver(){
        return this.lose() || this.win();
    }
    turn(reader){
        if(this.gameOver()){
            reader.close();
        }else{
            this.player.getMove(reader, pos => {
                if(!this.board.attack(pos)){
                    this.remainingMisses--;
                }
                this.board.print();
                console.log(this.remainingMisses);
                this.turn(reader);
            })
        }
    }
}

module.exports = Battleship;

File: AA_Exercise_Projects/W2D3/battleship_project/lib/battleship.rb
require_relative "board"
require_relative "player"

class Battleship
    attr_reader :board, :player
    def initialize(n)
        @player = Player.new
        @board = Board.new(n)
        @remaining_misses = @board.size / 2
    end
    def start_game
        @board.place_random_ships
        puts 'There are ' + @board.num_ships.to_s + ' ships on the board.'
        @board.print
    end
    def lose?
        if @remaining_misses <= 0
            puts 'you lose'
            true
        else
            false
        end
    end
    def win?
        if @board.num_ships == 0
            puts 'you win'
            true
        else
            false
        end
    end
    def game_over?
        self.lose? || self.win?
    end
    def turn
        move = @player.get_move
        if !@board.attack(move)
            @remaining_misses -= 1
        end
        @board.print
        puts 'remaining misses are:' + @remaining_misses.to_s
    end
end
